### PR TITLE
[exporter/clickhouse]Fix no such host when create database

### PR DESCRIPTION
**Pull request by Frapschen**
**Description:** Fix no such host when create database

error log:
```
Error: cannot build pipelines: failed to create "clickhouse" exporter, in pipeline "traces": cannot configure clickhouse traces exporter: create database:dial tcp: lookup my-default-clickhouse-headless on 169.254.25.10:53: no such host
```

in my config, the DSN is: `tcp://my-openinsight-clickhouse-headless:9000/openinsight` . it seem that `strings.Replace()` is not a good chooise:
https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/7f4d4425a03e7e47575211be489f912cd16ae509/exporter/clickhouseexporter/exporter_logs.go#L195

### CODE REVIEW

File: .chloggen/fix-clickhouse-create-database.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component:
```
Comment (JaredTan95) [inline .chloggen/fix-clickhouse-create-database.yaml:5]: Did you forget to add `component` belongs? :-P

**Review by JaredTan95 — CHANGES_REQUESTED** [review]

File: .chloggen/fix-clickhouse-create-database.yaml (hunk 1 above)
Comment (JaredTan95) [inline .chloggen/fix-clickhouse-create-database.yaml:5]:
```suggestion
component: clickhouseexporter
```

**Review by hanjm — APPROVED** [review]
LGTM.  https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/16091 will improve dsn to struct field

**Comment by hanjm** [comment]
hi @dmitryax @evan-bradley, could you help review, Thank you.

File: .chloggen/fix-clickhouse-create-database.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: clickhouseexporter
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix no such host when create database
```
Comment (evan-bradley) [inline .chloggen/fix-clickhouse-create-database.yaml:8]:
```suggestion
note: Fix "no such host" error when the DSN hostname contains the database name
```
Wording suggestion just to clarify the scope of the fix for users.

**Review by evan-bradley — APPROVED** [review]
Looks good to me.

**Comment by Frapschen** [comment]
@dmitryax can you review this pr?

**Comment by atoulme** [comment]
I think this fix would be best complemented with a unit test.

**Comment by Frapschen** [comment]
It's hard to test `createDatabase()` directly, so I use a part of `createDatabase()` codes to finish the test.

File: exporter/clickhouseexporter/exporter_logs_test.go [hunk 3]
```diff
@@ -101,6 +105,32 @@ func TestExporter_pushLogsData(t *testing.T) {
...
+	}
+
+	// a part of createDatabase()
+	createFunc := func(cfg *Config) string {
+		database, _ := parseDSNDatabase(cfg.DSN)
+		if database == defaultDatabase {
+			return ""
+		}
+		// use default database to create new database
+		dsnUseDefaultDatabase := fmt.Sprintf("%s%s", cfg.DSN[0:strings.LastIndex(cfg.DSN, database)], defaultDatabase)
```
Comment (atoulme) [inline exporter/clickhouseexporter/exporter_logs_test.go:127]: this doesn't actually call the method under test. Please make sure it does, as there is no guarantee your test does anything if the code changes.

**Comment by atoulme** [comment]
Please don't use gomonkey for this. Just isolate the function that changes the url and test it in isolation.

**Comment by Frapschen** [comment]
> Please don't use gomonkey for this. Just isolate the function that changes the url and test it in isolation.

ok, applied your suggestion.

**Comment by Frapschen** [comment]
@atoulme FIY

**Review by hanjm — APPROVED** [review]

File: exporter/clickhouseexporter/exporter_logs_test.go [hunk 4]
```diff
@@ -101,6 +102,38 @@ func TestExporter_pushLogsData(t *testing.T) {
 	})
 }
 
+func TestLogsExporter_createDatabase(t *testing.T) {
```
Comment (evan-bradley) [inline exporter/clickhouseexporter/exporter_logs_test.go:105]: Could you give more descriptive names to the tests in this function that explain what is being tested in each case? I also think having a case to test when the database name isn't a substring of the DSN would be helpful to make sure every case is covered.

File: exporter/clickhouseexporter/exporter_logs_test.go [hunk 5]
```diff
@@ -101,6 +102,38 @@ func TestExporter_pushLogsData(t *testing.T) {
...
+			RetrySettings:   exporterhelper.NewDefaultRetrySettings(),
+			DSN:             "tcp://mydatabase-clickhouse-headless:9000/mydatabase",
+			LogsTableName:   "otel_logs",
+			TracesTableName: "otel_traces",
+			TTLDays:         7,
+		}
+		err := createDatabase(c)
+		require.NoError(t, err)
+	})
+	t.Run("tcp://newdatabase-clickhouse-headless:9000/newdatabase", func(t *testing.T) {
```
Comment (evan-bradley) [inline exporter/clickhouseexporter/exporter_logs_test.go:122]: This appears to test the same scenario as the test above, is there a difference between the two outside of the database name?

**Review by evan-bradley — COMMENTED** [review]
Thanks for adding unit tests to cover this change.

**Comment by atoulme** [comment]
You didn't really isolate the line that changes the url with your unit test - you're creating a whole db. I would still just test that one line, but oh well. Like Evan said, this needs more tests - I'd like to see failures and no-op cases.

**Comment by JaredTan95** [comment]
@Frapschen  refactoring the line change into a func alone, and add UT for it. that's better, I think.

File: exporter/clickhouseexporter/exporter_logs_test.go (hunk 4 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_logs_test.go]: done

File: exporter/clickhouseexporter/exporter_logs_test.go (hunk 5 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_logs_test.go]: done

**Comment by Frapschen** [comment]
> You didn't really isolate the line that changes the url with your unit test - you're creating a whole db. I would still just test that one line, but oh well. Like Evan said, this needs more tests - I'd like to see failures and no-op cases.

@atoulme your advice are applied. please review :)

File: exporter/clickhouseexporter/exporter_logs.go [hunk 6]
```diff
@@ -192,7 +192,7 @@ func createDatabase(cfg *Config) error {
 		return nil
 	}
 	// use default database to create new database
-	dsnUseDefaultDatabase := strings.Replace(cfg.DSN, database, defaultDatabase, 1)
+	dsnUseDefaultDatabase, _ := getDefaultDns(cfg.DSN, database)
```
Comment (atoulme) [inline exporter/clickhouseexporter/exporter_logs.go:195]: can you do something with the error returned?

File: exporter/clickhouseexporter/exporter_logs_test.go [hunk 7]
```diff
@@ -101,6 +101,29 @@ func TestExporter_pushLogsData(t *testing.T) {
...
+		dsn := "tcp://newdatabase-clickhouse-headless:9000/otel"
+		defaultDns, _ := getDefaultDns(dsn, "otel")
+		require.Equal(t, defaultDns, "tcp://newdatabase-clickhouse-headless:9000/default")
+	})
+	t.Run("error param for database", func(t *testing.T) {
+		dsn := "tcp://mydatabase-clickhouse-headless:9000/mydatabase"
+		_, err := getDefaultDns(dsn, "otel")
+		require.Error(t, err)
+	})
+	t.Run("database name is same as default default database", func(t *testing.T) {
```
Comment (atoulme) [inline exporter/clickhouseexporter/exporter_logs_test.go:120]:
```suggestion
	t.Run("database name is same as default database", func(t *testing.T) {
```

File: exporter/clickhouseexporter/exporter_logs.go (hunk 6 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_logs.go]: done

File: exporter/clickhouseexporter/exporter_logs_test.go (hunk 7 above)
Comment (Frapschen) [reply exporter/clickhouseexporter/exporter_logs_test.go]: done

**Review by atoulme — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

**Pull request merged**